dbt/adapters/duckdb/relation.py

```diff
@@ -0,0 +1,43 @@
+from dataclasses import dataclass
+from typing import Any
+from typing import Optional
+from typing import Type
+
+from dbt.adapters.base.relation import BaseRelation
+from dbt.adapters.base.relation import Self
+from dbt.contracts.graph.parsed import ParsedSourceDefinition
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class DuckDBRelation(BaseRelation):
+    external_location: Optional[str] = None
+
+    @classmethod
+    def create_from_source(cls: Type[Self], source: ParsedSourceDefinition, **kwargs: Any) -> Self:
+
+        # Some special handling here to allow sources that are external files to be specified
+        # via a `external_location` meta field. If the source's meta field is used, we include
+        # some logic to allow basic templating of the external location based on the individual
+        # name or identifier for the table itself to cut down on boilerplate.
+        ext_location = None
+        if "external_location" in source.meta:
+            ext_location = source.meta["external_location"]
+        elif "external_location" in source.source_meta:
+            # Use str.format here to allow for some basic templating outside of Jinja
+            ext_location = source.source_meta["external_location"].format(
```

Comment:
Should not we allow templating in both `source.meta` and `source.source_meta`?

Reply:
I started that direction at first, but then I wasn't sure; if you're specifying a location in `source.meta`, then you already know the name of the table, so I wasn't sure how the templating helped-- are you thinking it's just better for consistency purposes, even though it's not as helpful when it's on `source.meta`?

Reply:
I was more thinking about consistency. But I guess this makes sense too

Reply:
Or more precisely that source.meta is setting common for all underlying sources.

Reply:
So the naming is confusing here-- the `source` refers to the node (aka the table/view/file) itself, which is allowed to have its own `meta` dictionary, as well as the top-level `sources:` object in the yaml (which includes a list of `tables`) that is _also_ allowed to have its own `meta` dictionary.

But leaving all of that aside, your point about consistency is a good one b/c there's no harm in enforcing it here, so I will make a change to that effect. I'm also going to add in some tests I derive from https://github.com/dbt-labs/dbt-core/tree/main/tests/functional/sources and will then merge it and, dare I say it, cut a release?!

Reply:
Nothing else you want to get in to 1.3.0, right?